Added JSON Validation, upstream HTTP timeout / exceptions, and additional logging

telnetdoogie opened the pull request:
Should solve #19
This seems to respond quickly, it also prevents multiple clients / threads triggering a fresh download from the parent server as well.
I believe #19 was caused by some invalid JSON (should now be logged appropriately) from the source.

telnetdoogie committed:
Added Caching, JSON Validation, and additional logging
Should solve #19

telnetdoogie committed:
Fixed erroneous '$' on ouput

matthuisman commented:
this is getting overly complicated.

adding locks is just asking for trouble

If we want cache, there are some python libs that can do that for us.

If the json data was bad, it would error before any headers are sent and therefore not be a broken pipe.

But i agree there is something after sending initial headers and finishing the response that is raising

Please test with your changes and find what is actually causing the error, we can then fix the root cause

telnetdoogie committed:
Removed inadvertently re-added CHUCNKSIZE attr from my foolish misfork

telnetdoogie commented:
The error no longer occurs. The LOCK merely makes a 2nd request wait to enter the method until the first caller has completed, and in that case the cached response it returned.

telnetdoogie commented:
@matthuisman I can remove the custom in-memory caching and use TTLCache from cachetools if you'd prefer?

matthuisman commented:
"The error no longer occurs with this."

So what was the root cause?

telnetdoogie commented:
the broken pipe error was occuring when client times out, which is why I added the try catch blocks.
In order to reduce the time to generate the responses, I also introduced the caching, in order to minimize the opportunity for client timeouts.

matthuisman commented:
the broken pipe can happen when there is an error in code after sending headers which I believe could be the case here

matthuisman commented:
this wasnt an issue before the recent changes, so it being a timeout on the app.json which is now actually smaller doesnt make sense to me?

telnetdoogie commented:
multiple / concurrent requests would originally download the epg AND unzip - and that's quite a large payload, meaning on a device like mine, an older synology running docker (which isn't particularly quick to begin with) the timeouts are much more likely...

matthuisman commented:
so we should just go back to providing channels the epg .gz ? 
I still didnt understand why that was causing issues:
https://github.com/matthuisman/samsung-tvplus-for-channels/pull/18#issuecomment-2455660436

matthuisman commented:
Or allow both

telnetdoogie commented:
it's not 100% reproducible... perhaps when the upstream is not responding quickly it further exacerbates the issue. 
Regardless, it probably doesn't make sense for every request to hit the upstream nor to unzip the epg.
Whether you unzip or serve zipped it would still make sense to limit the upstream requests by caching, no?

telnetdoogie commented:
Perhaps the issues this morning were a problem with the upstream entirely and not a problem with the code in this repo

matthuisman commented:
i agree on caching but i expected everyone using these containers would control how often they are hit and have the upstream of these taking care of how often they call it rather than using extra memory caching.

If we want to do caching, I reckon we use: https://cachelib.readthedocs.io/en/stable/file/#cachelib.file.FileSystemCache

But, i was trying to not have external dependencies for users who want to use python file directly :(
If we going away from that, we may as well move to something like Flask

telnetdoogie committed:
Added BrokenPipeError catch for EPG data

telnetdoogie committed:
removed extraneous colon on BrokenPipe Catch output

telnetdoogie commented:
I was able to reproduce the original error by calling `/epg.xml?regions=us` when not cached, and  closing the client before the response is sent.

What was:
```log

192.168.1.6 - - [10/Nov/2024 17:40:55] "GET /epg.xml?regions=us HTTP/1.1" 500 -
----------------------------------------
Exception occurred during processing of request from ('192.168.1.6', 53582)
Traceback (most recent call last):
  File "/usr/src/app/./app.py", line 54, in do_GET
    routes[func]()
  File "/usr/src/app/./app.py", line 140, in _epg
    self.wfile.write(chunk)
  File "/usr/local/lib/python3.12/socketserver.py", line 840, in write
    self._sock.sendall(b)
BrokenPipeError: [Errno 32] Broken pipe
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/socketserver.py", line 692, in process_request_thread
    self.finish_request(request, client_address)
  File "/usr/local/lib/python3.12/socketserver.py", line 362, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/src/app/./app.py", line 24, in __init__
    super().__init__(*args, **kwargs)
  File "/usr/local/lib/python3.12/socketserver.py", line 761, in __init__
    self.handle()
  File "/usr/local/lib/python3.12/http/server.py", line 436, in handle
    self.handle_one_request()
  File "/usr/local/lib/python3.12/http/server.py", line 424, in handle_one_request
    method()
  File "/usr/src/app/./app.py", line 56, in do_GET
    self._error(e)
  File "/usr/src/app/./app.py", line 28, in _error
    self.end_headers()
  File "/usr/local/lib/python3.12/http/server.py", line 538, in end_headers
    self.flush_headers()
  File "/usr/local/lib/python3.12/http/server.py", line 542, in flush_headers
    self.wfile.write(b"".join(self._headers_buffer))
  File "/usr/local/lib/python3.12/socketserver.py", line 840, in write
    self._sock.sendall(b)
BrokenPipeError: [Errno 32] Broken pipe
----------------------------------------
```

Now looks like:
```log
Received request on /epg.xml with params: {'regions': 'us'}
Using cached EPG data.
192.168.1.6 - - [10/Nov/2024 17:38:01] "GET /epg.xml?regions=us HTTP/1.1" 200 -
Client disconnected while sending EPG data
```

telnetdoogie commented:
Would you like me to remove the caching portion if you're uncomfortable with it and leave the try catches for Broken Pipe in place and the JSON check / alerts?

Let me know how you'd like me to proceed. 

I can:
* keep as is
* remove caching, keep try/catch and JSON validation
* remove caching & try/catch, keep  JSON validation
* close / abandon PR

matthuisman commented:
your comment before said you still got closed pip with a cached file?

telnetdoogie commented:
I got it uncached. I'm sure You can still get it with a cached file but the opportunity is much smaller and harder to reproduce (you'd have to be much quicker, I didn't do it programmatically) 

The root problem I believe based on that reproduction is when the upstream connection takes a while and the client disconnects (maybe due to a client timeout)

telnetdoogie commented:
just a few minutes ago for instance, i.mjh.nz was not responding. It wasn't failing, no errors, just not responding... When the upstream is doing this, there's no timeout in this app and any client will likely timeout before the app does, so when it eventually resolves, the closed pipe issue will happen.

telnetdoogie commented:
so... forget caching. It doesn't help

telnetdoogie commented:
I will at a minimum remove the caching and related locks

matthuisman commented:
i.mjh.nz is all behind cloudflare cache.
I dont expect them to have that many issues..?

telnetdoogie commented:
Not sure what's happening in cloudflare... in the earlier situation, as the app requested the xml, it sat doing nothing. I opened a browser and opened the URL there, it too sat there... it connected, but no response was returned. No 500, no 200, no 400... just waiting.

matthuisman commented:
maybe its the server itself and not requests?
Do we know its pausing at the requests download?
If you log before and after, should be able to tell

matthuisman commented:
might be a CF issue. I'm having very slow access to https://i.mjh.nz/SamsungTVPlus/.channels.json right now

matthuisman commented:
and https://i.mjh.nz/SamsungTVPlus/.channels.json.gz very slow download

matthuisman commented:
I think us switching to use .json.gz would also be good improvement.
Its quite a bit smaller than just the .json

telnetdoogie committed:
Reverting to 9ee43f5

telnetdoogie committed:
Added upstream timeouts

telnetdoogie committed:
catch BrokenPipe

telnetdoogie committed:
move catch for BrokenPipe

telnetdoogie committed:
other HTTP Exceptions included

telnetdoogie committed:
Testing timeout

telnetdoogie committed:
Testing Error message in log

telnetdoogie committed:
Testing completed, error passed to log

telnetdoogie commented:
Sorry for all the commit noise... definitely will want to squash this one if you take it.
Reverted to 9ee43f5, and added:
* try/catch for requests upstream
* added connect and complete timeouts for upstream requests
* added try/catch for Broken Pipe (client disconnects before response is completed)
* expect valid JSON before continuing (throw exception and log if not)

telnetdoogie commented:
A recommended 'update' to this, which I did not make... could be a feature request... would be to store successful / valid payloads in memory, and in the case of a failed download / timeout, if previous valid payload is still in memory, use the 'stale' version instead from memory. Not quite caching but prevention from complete failed response at least.
Downside would be that clients would not receive errors and might not know if something has gone wrong.

telnetdoogie committed:
Removed duplicate log output

telnetdoogie committed:
Extracted method from _status and _playlist; remove duplication

matthuisman commented:
@telnetdoogie thanks for your work on this and helping debug :)

For now, Ive just pushed a few changes to master including your timeout change which is a good idea!
Also moved to use the .gz for json as quite a bit smaller.
Support for using the smaller XML's if using a single region
And a easier to follow status page IMO

Please keep this PR open as we discuss / work through other things
